bpf/kmesh/ads/include/circuit_breaker.h

```diff
@@ -0,0 +1,168 @@
+/* SPDX-License-Identifier: (GPL-2.0-only OR BSD-2-Clause) */
+/* Copyright Authors of Kmesh */
+
+#include "bpf_log.h"
+#include "kmesh_common.h"
+#include "bpf_common.h"
+
+#ifndef __KMESH_CIRCUIT_BREAKER_H__
+#define __KMESH_CIRCUIT_BREAKER_H__
+
+#define CLUSTER_NAME_MAX_LEN BPF_DATA_MAX_LEN
+
+struct cluster_stats {
+    __u32 active_connections;
+};
+
+struct cluster_stats_key {
+    __u64 netns_cookie;
+    __u32 cluster_id;
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_HASH);
+    __uint(key_size, sizeof(struct cluster_stats_key));
+    __uint(value_size, sizeof(struct cluster_stats));
+    __uint(map_flags, BPF_F_NO_PREALLOC);
+    __uint(max_entries, MAP_SIZE_OF_CLUSTER);
+} map_of_cluster_stats SEC(".maps");
+
+struct cluster_sock_data {
+    __u32 cluster_id;
+};
+
+struct {
+    __uint(type, BPF_MAP_TYPE_SK_STORAGE);
+    __uint(map_flags, BPF_F_NO_PREALLOC);
+    __type(key, int);
+    __type(value, struct cluster_sock_data);
+} map_of_cluster_sock SEC(".maps");
+
+static inline void update_cluster_active_connections(const struct cluster_stats_key *key, int delta)
+{
+    if (!key) {
+        return;
+    }
+    struct cluster_stats *stats = NULL;
+    stats = kmesh_map_lookup_elem(&map_of_cluster_stats, key);
+    if (!stats) {
+        struct cluster_stats new_stats = {0};
+        new_stats.active_connections = delta;
+        BPF_LOG(
+            DEBUG,
+            KMESH,
+            "create new stats(netns_cookie = %lld, cluster_id = %ld)",
+            key->netns_cookie,
+            key->cluster_id);
+        kmesh_map_update_elem(&map_of_cluster_stats, key, &new_stats);
+    } else {
+        stats->active_connections += delta;
+        kmesh_map_update_elem(&map_of_cluster_stats, key, stats);
+        BPF_LOG(
+            DEBUG,
+            KMESH,
+            "update existing stats(netns_cookie = %lld, cluster_id = %ld)",
+            key->netns_cookie,
+            key->cluster_id);
+    }
+}
+
+static inline int on_cluster_sock_bind(ctx_buff_t *ctx, const Cluster__Cluster *cluster)
+{
+    __u32 cluster_id = cluster->id;
+    struct cluster_stats_key key = {0};
+    __u64 cookie = bpf_get_netns_cookie(ctx);
+    key.cluster_id = cluster_id;
+    key.netns_cookie = cookie;
+    struct cluster_stats *stats = NULL;
+    stats = kmesh_map_lookup_elem(&map_of_cluster_stats, &key);
+
+    if (stats != NULL) {
+        Cluster__CircuitBreakers *cbs = NULL;
+        cbs = kmesh_get_ptr_val(cluster->circuit_breakers);
+        if (cbs != NULL && stats->active_connections >= cbs->max_connections) {
+            BPF_LOG(
+                DEBUG,
+                KMESH,
+                "Current active connections %d exceeded max connections %d, reject connection\n",
+                stats->active_connections,
+                cbs->max_connections);
+            return 0;
```

Comment:
We should always use 0 as success and negative number as error code

Reply:
ok